// thinkjs/lib/adapter/template/nunjucks.js
'use strict';
/**
 * nunjucks template
 * @type {Class}
 */

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

exports.__esModule = true;

var _default = (function (_think$adapter$template) {
  _inherits(_default, _think$adapter$template);

  function _default() {
    _classCallCheck(this, _default);

    _think$adapter$template.apply(this, arguments);
  }

  /**
   * run
   * @param  {String} templateFile []
   * @param  {Object} tVar         []
   * @return {Promise}             []
   */

  _default.prototype.run = function run(templateFile, tVar, config) {
    var nunjucks, conf, env, prerender;
    return _regeneratorRuntime.async(function run$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          context$2$0.next = 2;
          return _regeneratorRuntime.awrap(think.npm('nunjucks'));

        case 2:
          nunjucks = context$2$0.sent;
          conf = think.extend({
            autoescape: false,
            watch: false,
            noCache: false,
            throwOnUndefined: false
          }, think.config('view.options'), config && config.options);
          env = nunjucks.configure(think.ROOT_PATH, conf);
          prerender = config && config.prerender;

          prerender = prerender || think.config('view.prerender');
          if (think.isFunction(prerender)) {
            prerender(nunjucks, env);
          }

          return context$2$0.abrupt('return', nunjucks.render(templateFile, tVar));

        case 9:
        case 'end':
          return context$2$0.stop();
      }
    }, null, this);
  };

  return _default;
})(think.adapter.template);

exports['default'] = _default;
module.exports = exports['default'];

//pre render
